src/EventStore.Projections.Core/Services/Processing/CoreProjectionCheckpointManager.cs

```diff
@@ -175,6 +177,17 @@ public abstract class CoreProjectionCheckpointManager : IProjectionCheckpointMan
 
 			if (partition == "")
 				_currentProjectionState = newState;
+			CheckStateSize(_currentProjectionState);
+		}
+
+		private void CheckStateSize(PartitionState result) {
```

Comment:
You could actually do the partition check here as well instead of in `PartitionStateUpdateManager`. If you move the `if (partition == "" && newState.State == null)` condition, you could check the state size of `newState` before the other two conditions. e.g:

```
if (partition == "" && newState.State == null) // ignore non-root partitions and non-changed states
	throw new NotSupportedException("Internal check");

CheckStateSize(newState, partition);

if (_usePersistentCheckpoints && partition != "")
	CapturePartitionStateUpdated(partition, oldState, newState);

if (partition == "")
	_currentProjectionState = newState;
```

Reply:
Done